packages/jest-runtime/src/transform.js

```diff
@@ -276,21 +280,55 @@ const transformSource = (
     return result;
   }
 
-  result = content;
+  let transformed: TransformedSource = {
+    content,
+    sourceMap: null,
+  };
 
   if (transform && shouldTransform(filename, config)) {
-    result = transform.process(result, filename, config, {
+    const processed = transform.process(content, filename, config, {
       instrument,
       watch: config.watch,
     });
+
+    if (typeof processed === 'string') {
+      transformed.content = processed;
+    } else {
+      transformed = processed;
+    }
+  }
+
+  if (config.mapCoverage) {
+    if (!transformed.sourceMap) {
+      const convert = require('convert-source-map');
+      const inlineSourceMap = convert.fromSource(transformed.content);
+      if (inlineSourceMap) {
+        transformed.sourceMap = inlineSourceMap.toJSON();
+      }
+    }
+  } else {
+    transformed.sourceMap = null;
   }
 
   // That means that the transform has a custom instrumentation
   // logic and will handle it based on `config.collectCoverage` option
-  const transformWillInstrument = transform && transform.canInstrument;
+  const transformDidInstrument = transform && transform.canInstrument;
+
+  if (!transformDidInstrument && instrument) {
+    result = instrumentFile(transformed.content, filename, config);
+  } else {
+    result = transformed.content;
+  }
 
-  if (!transformWillInstrument && instrument) {
-    result = instrumentFile(result, filename, config);
+  if (instrument && transformed.sourceMap && config.mapCoverage) {
+    const sourceMapContent = typeof transformed.sourceMap === 'string'
+      ? transformed.sourceMap
+      : JSON.stringify(transformed.sourceMap);
+    const sourceMapFilePath = cacheFilePath + '.map';
+    writeCacheFile(sourceMapFilePath, sourceMapContent);
+    result +=
+      `\n;global.__coverage__['${escapePathForJavaScript(filename)}']` + 
```

Comment:
@dmitriiabramov well in practice, it can pretty much be assumed that `__covearge__` is the variable (I don't know that I've seen many deviations), this variable can be configured:

https://github.com/istanbuljs/istanbul-lib-instrument/blob/b08e4f55d60a43d542add9b09b3bf52a1d88ce6f/src/instrumenter.js#L13